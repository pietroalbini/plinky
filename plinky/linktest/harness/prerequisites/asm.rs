use crate::tests::{TestArch, TestExecution};
use crate::utils::run;
use anyhow::{anyhow, Error};
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct AsmFile {
    source: String,
    format: Option<AsmFormat>,
    output: Option<String>,
}

impl AsmFile {
    pub(super) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        let source = execution
            .test
            .files
            .get(&*self.source)
            .ok_or_else(|| anyhow!("missing {}", self.source))?;

        let dest_name = match &self.output {
            Some(output) => output.clone(),
            None => format!(
                "{}.o",
                self.source.rsplit_once('.').map(|(name, _ext)| name).unwrap_or(&self.source)
            ),
        };

        let source_dir = TempDir::new()?;
        std::fs::write(source_dir.path().join(&self.source), *source)?;

        eprintln!("compiling {} into {dest_name}...", self.source);
        run(Command::new("nasm")
            .current_dir(source_dir.path())
            .arg("-f")
            .arg(match (&self.format, execution.arch) {
                (Some(AsmFormat::Elf32), _) | (None, TestArch::X86) => "elf32",
                (Some(AsmFormat::Elf64), _) | (None, TestArch::X86_64) => "elf64",
            })
            .arg("-o")
            .arg(dest_dir.join(dest_name))
            .arg(&self.source))?;

        Ok(())
    }
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
enum AsmFormat {
    Elf32,
    Elf64,
}
