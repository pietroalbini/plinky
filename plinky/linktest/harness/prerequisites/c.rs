use crate::tests::{TestArch, TestExecution};
use crate::utils::run;
use anyhow::Error;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

#[derive(serde::Deserialize, Clone)]
#[serde(deny_unknown_fields)]
pub(super) struct CFile {
    source: String,
    output: Option<String>,
    libc: Libc,
    relocation: Relocation,
}

impl CFile {
    pub(super) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        let source = execution.file(&self.source)?;

        let dest_name = match &self.output {
            Some(output) => output.clone(),
            None => format!(
                "{}.o",
                self.source.rsplit_once('.').map(|(name, _ext)| name).unwrap_or(&self.source)
            ),
        };

        let source_dir = TempDir::new()?;
        std::fs::write(source_dir.path().join(&self.source), source)?;

        eprintln!("compiling {} into {dest_name}...", self.source);
        run(Command::new("cc")
            .current_dir(source_dir.path())
            .arg("-c")
            // Disable control-flow protection, as it's not implemented in the linker right now.
            .arg("-fcf-protection=none")
            .arg("-o")
            .arg(dest_dir.join(dest_name))
            .arg(match execution.arch {
                TestArch::X86 => "-m32",
                TestArch::X86_64 => "-m64",
            })
            .args(match self.libc {
                Libc::Freestanding => &["-nostdlib"],
            })
            .args(match self.relocation {
                Relocation::Static => &["-fno-pic"] as &[&str],
                Relocation::PicOnlyGot => &["-fPIC", "-fno-plt"],
            })
            .arg(&self.source))?;

        Ok(())
    }
}

#[derive(serde::Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
enum Libc {
    Freestanding,
}

#[derive(serde::Deserialize, Clone)]
#[serde(rename_all = "kebab-case")]
enum Relocation {
    Static,
    PicOnlyGot,
}
