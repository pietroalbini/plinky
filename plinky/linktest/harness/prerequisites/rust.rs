use crate::tests::TestExecution;
use crate::utils::run;
use anyhow::{anyhow, Error};
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct RustFile {
    source: String,
    #[serde(default)]
    panic: Panic,
}

impl RustFile {
    pub(super) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        let source = execution
            .test
            .files
            .get(&*self.source)
            .ok_or_else(|| anyhow!("missing {}", self.source))?;

        let dest_name = format!(
            "lib{}.a",
            self.source.rsplit_once('.').map(|(name, _ext)| name).unwrap_or(&self.source)
        );

        let source_dir = TempDir::new()?;
        std::fs::write(source_dir.path().join(&self.source), *source)?;

        run(Command::new("rustc")
            .current_dir(source_dir.path())
            .arg(&self.source)
            .arg("-o")
            .arg(dest_dir.join(dest_name))
            .arg("--crate-type=staticlib")
            .arg(match self.panic {
                Panic::Abort => "-Cpanic=abort",
            }))?;

        Ok(())
    }
}

#[derive(serde::Deserialize, Default)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
enum Panic {
    #[default]
    Abort,
}
