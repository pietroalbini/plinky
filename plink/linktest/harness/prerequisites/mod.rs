mod asm;
mod c;

use crate::prerequisites::asm::AsmFile;
use crate::prerequisites::c::CFile;
use crate::TestExecution;
use anyhow::Error;
use std::path::Path;

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct Prerequisites {
    #[serde(default)]
    asm: Vec<AsmFile>,
    #[serde(default)]
    c: Vec<CFile>,
}

impl Prerequisites {
    pub(crate) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        for asm in &self.asm {
            asm.build(execution, dest_dir)?;
        }
        for c in &self.c {
            c.build(execution, dest_dir)?;
        }
        Ok(())
    }
}
