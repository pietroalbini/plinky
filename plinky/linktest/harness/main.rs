mod prerequisites;
mod utils;

use crate::prerequisites::Prerequisites;
use anyhow::{anyhow, bail, Error};
use std::collections::HashMap;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

struct Test {
    name: &'static str,
    files: HashMap<&'static str, &'static [u8]>,
}

impl Test {
    fn run(self) -> Result<(), Error> {
        let settings: TestSettings = toml::from_str(std::str::from_utf8(
            self.files.get("test.toml").ok_or_else(|| anyhow!("missing test.toml"))?,
        )?)?;

        for arch in &settings.archs {
            let root = TempDir::new()?;
            TestExecution {
                test: &self,
                settings: &settings,
                root: root.path(),
                suffix: match arch {
                    TestArch::X86 => "-32bit",
                    TestArch::X86_64 => "-64bit",
                },
                arch: *arch,
            }
            .run()?;
        }

        Ok(())
    }
}

struct TestExecution<'a> {
    test: &'a Test,
    settings: &'a TestSettings,
    root: &'a Path,
    suffix: &'a str,
    arch: TestArch,
}

impl TestExecution<'_> {
    fn run(self) -> Result<(), Error> {
        self.settings.prerequisites.build(&self, self.root)?;
        match self.settings.kind {
            TestKind::LinkFail => self.run_link_fail(),
            TestKind::LinkPass => self.run_link_pass(),
            TestKind::RunFail => self.run_run_fail(),
            TestKind::RunPass => self.run_run_pass(),
        }
    }

    fn run_link_fail(self) -> Result<(), Error> {
        if self.link_and_snapshot()? {
            bail!("linking was supposed to fail but passed!");
        }
        Ok(())
    }

    fn run_link_pass(self) -> Result<(), Error> {
        if !self.link_and_snapshot()? {
            bail!("linking was supposed to pass but failed!");
        }
        Ok(())
    }

    fn run_run_fail(self) -> Result<(), Error> {
        if self.run_and_snapshot()? {
            bail!("running was supposed to fail but passed!");
        }
        Ok(())
    }

    fn run_run_pass(self) -> Result<(), Error> {
        if !self.run_and_snapshot()? {
            bail!("running was supposed to pass but failed!");
        }
        Ok(())
    }

    fn link_and_snapshot(&self) -> Result<bool, Error> {
        let mut command = Command::new(env!("CARGO_BIN_EXE_plinky"));
        command.current_dir(self.root).args(&self.settings.cmd).env("RUST_BACKTRACE", "1");
        for debug_print in &self.settings.debug_print {
            command.args(["--debug-print", debug_print]);
        }

        self.record_snapshot("linker", "linking", &mut command)
    }

    fn run_and_snapshot(&self) -> Result<bool, Error> {
        if !self.link_and_snapshot()? {
            bail!("linking was supposed to pass but failed!");
        }

        let mut command = Command::new(self.root.join("a.out"));
        command.current_dir(self.root);

        self.record_snapshot("run", "running", &mut command)
    }

    fn record_snapshot(
        &self,
        snapshot_name: &str,
        action: &str,
        command: &mut Command,
    ) -> Result<bool, Error> {
        let snapshot_name = format!("{snapshot_name}{}", self.suffix);

        let (output_repr, success) = match command.output() {
            Ok(output) => {
                let mut output_repr = format!("{action} exited with {}\n", output.status);
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.is_empty() {
                        output_repr.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let content = String::from_utf8_lossy(content);
                        let content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        output_repr.push_str(&format!("\n=== {name} ===\n{}\n", content,));
                    }
                }
                (output_repr, output.status.success())
            }
            Err(err) => (format!("{action} failed to execute with error: {err}"), false),
        };

        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(std::fs::canonicalize(
            Path::new(env!("CARGO_MANIFEST_DIR"))
                .join("linktest")
                .join("tests")
                .join(self.test.name),
        )?);

        insta_settings.bind(|| {
            insta::assert_snapshot!(snapshot_name, output_repr);
        });
        Ok(success)
    }
}

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
struct TestSettings {
    cmd: Vec<String>,
    kind: TestKind,
    #[serde(default = "default_test_archs")]
    archs: Vec<TestArch>,
    #[serde(default)]
    debug_print: Vec<String>,
    #[serde(flatten)]
    prerequisites: Prerequisites,
}

#[derive(serde::Deserialize, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
enum TestArch {
    X86,
    X86_64,
}

fn default_test_archs() -> Vec<TestArch> {
    vec![TestArch::X86_64]
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
enum TestKind {
    LinkFail,
    LinkPass,
    RunFail,
    RunPass,
}

macro_rules! linktest {
    ($name:ident, files[$($file:expr),*]) => {
        #[test]
        fn $name() {
            let mut files = HashMap::new();
            $(files.insert($file.rsplit_once('/').unwrap().1, include_bytes!($file) as &[u8]);)*
            Test {
                name: stringify!($name),
                files,
            }.run().unwrap();
        }
    }
}

include!(concat!(env!("OUT_DIR"), "/linktest_definition.rs"));
