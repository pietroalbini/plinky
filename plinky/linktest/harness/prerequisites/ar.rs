use crate::prerequisites::Prerequisites;
use crate::tests::TestExecution;
use crate::utils::run;
use anyhow::Error;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub(super) struct ArArchive {
    name: String,
    #[serde(default)]
    symbol_table: Option<bool>,
    #[serde(flatten)]
    content: Prerequisites,
}

impl ArArchive {
    pub(super) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        let inputs_dir = TempDir::new()?;
        self.content.build(execution, inputs_dir.path())?;

        let mut to_archive = Vec::new();
        for entry in std::fs::read_dir(inputs_dir.path())? {
            let path = entry?.path();
            if path.is_file() {
                to_archive.push(path.file_name().unwrap().to_os_string());
            }
        }
        to_archive.sort();

        let mut flags = "rc".to_string();
        match self.symbol_table {
            None | Some(true) => flags.push('s'),
            Some(false) => flags.push('S'),
        }

        println!("archiving {to_archive:?} into {}...", self.name);
        run(Command::new("ar")
            .current_dir(inputs_dir.path())
            .arg(flags)
            .arg(dest_dir.join(&self.name))
            .args(&to_archive))?;
        Ok(())
    }
}
