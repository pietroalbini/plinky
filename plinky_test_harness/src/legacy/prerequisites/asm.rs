use crate::legacy::prerequisites::Arch;
use crate::utils::run;
use anyhow::Error;
use std::path::Path;
use std::process::Command;

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(deny_unknown_fields)]
pub(super) struct AsmFile {
    source: String,
    format: Option<AsmFormat>,
    output: Option<String>,
}

impl AsmFile {
    pub(super) fn build(
        &self,
        arch: Arch,
        source_dir: &Path,
        dest_dir: &Path,
    ) -> Result<(), Error> {
        let dest_name = match &self.output {
            Some(output) => output.clone(),
            None => format!(
                "{}.o",
                self.source.rsplit_once('.').map(|(name, _ext)| name).unwrap_or(&self.source)
            ),
        };

        eprintln!("compiling {} into {dest_name}...", self.source);
        run(Command::new("as")
            .current_dir(source_dir)
            .arg(match (&self.format, arch) {
                (Some(AsmFormat::Elf32), _) | (None, Arch::X86) => "--32",
                (Some(AsmFormat::Elf64), _) | (None, Arch::X86_64) => "--64",
            })
            .arg("-o")
            .arg(dest_dir.join(dest_name))
            .arg(&self.source))?;

        Ok(())
    }
}

#[derive(serde::Deserialize, Debug, Clone, Copy)]
#[serde(rename_all = "kebab-case")]
enum AsmFormat {
    Elf32,
    Elf64,
}
