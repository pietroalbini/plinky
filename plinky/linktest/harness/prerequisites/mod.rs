mod ar;
mod asm;
mod c;

use crate::prerequisites::ar::ArArchive;
use crate::prerequisites::asm::AsmFile;
use crate::prerequisites::c::CFile;
use crate::{TestArch, TestExecution};
use anyhow::Error;
use std::collections::BTreeMap;
use std::path::Path;

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct Prerequisites {
    #[serde(default)]
    asm: Vec<AsmFile>,
    #[serde(default)]
    c: Vec<CFile>,
    #[serde(default)]
    ar: Vec<ArArchive>,
    #[serde(default)]
    arch: BTreeMap<TestArch, Prerequisites>,
}

impl Prerequisites {
    pub(crate) fn build(&self, execution: &TestExecution, dest_dir: &Path) -> Result<(), Error> {
        for asm in &self.asm {
            asm.build(execution, dest_dir)?;
        }
        for c in &self.c {
            c.build(execution, dest_dir)?;
        }
        for ar in &self.ar {
            ar.build(execution, dest_dir)?;
        }
        if let Some(arch_specific) = self.arch.get(&execution.arch) {
            arch_specific.build(execution, dest_dir)?;
        }
        Ok(())
    }
}
