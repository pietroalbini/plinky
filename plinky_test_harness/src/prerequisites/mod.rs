mod ar;
mod asm;
mod c;
mod ld;
mod rust;

use crate::prerequisites::ar::ArArchive;
use crate::prerequisites::asm::AsmFile;
use crate::prerequisites::c::CFile;
use crate::prerequisites::ld::LdInvocation;
use crate::prerequisites::rust::RustFile;
use anyhow::Error;
use std::collections::BTreeMap;
use std::path::Path;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, serde::Deserialize)]
pub enum Arch {
    #[serde(rename = "x86")]
    X86,
    #[serde(rename = "x86_64")]
    X86_64,
}

#[derive(serde::Deserialize, Debug, Clone)]
#[serde(deny_unknown_fields)]
pub struct Prerequisites {
    #[serde(default)]
    asm: Vec<AsmFile>,
    #[serde(default)]
    c: Vec<CFile>,
    #[serde(default)]
    ar: Vec<ArArchive>,
    #[serde(default)]
    ld: Vec<LdInvocation>,
    #[serde(default)]
    rust: Vec<RustFile>,
    #[serde(default)]
    arch: BTreeMap<Arch, Prerequisites>,
}

impl Prerequisites {
    pub fn build(&self, arch: Arch, source_dir: &Path, dest_dir: &Path) -> Result<(), Error> {
        for asm in &self.asm {
            asm.build(arch, source_dir, dest_dir)?;
        }
        for c in &self.c {
            c.build(arch, source_dir, dest_dir)?;
        }
        for ar in &self.ar {
            ar.build(arch, source_dir, dest_dir)?;
        }
        for rust in &self.rust {
            rust.build(arch, source_dir, dest_dir)?;
        }
        for ld in &self.ld {
            ld.build(arch, source_dir, dest_dir)?;
        }
        if let Some(arch_specific) = self.arch.get(&arch) {
            arch_specific.build(arch, source_dir, dest_dir)?;
        }
        Ok(())
    }
}
