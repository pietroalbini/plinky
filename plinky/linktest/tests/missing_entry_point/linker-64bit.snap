---
source: plinky/linktest/harness/main.rs
---
linking exited with exit status: 1

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  env: ElfEnvironment {
 │      class: Elf64,
 │      endian: Little,
 │      abi: SystemV,
 │      machine: X86_64,
 │  }
 │
 │  section .text(section#1) in hello.o
 │   │
 │   │  ╭────┬───╮
 │   │  │ 90 │ . │
 │   │  ╰────┴───╯
 │   ┴
 │
 │    Symbols:
 │  ╭──────────────┬──────────┬────────────┬────────────────────────╮
 │  │ Name         │ Source   │ Visibility │ Value                  │
 │  ├──────────────┼──────────┼────────────┼────────────────────────┤
 │  │              │ hello.o  │ local      │ <undefined>            │
 │  ├──────────────┼──────────┼────────────┼────────────────────────┤
 │  │              │ hello.o  │ local      │ .text(section#1) + 0x0 │
 │  ├──────────────┼──────────┼────────────┼────────────────────────┤
 │  │ _start       │ <plinky> │ global     │ <undefined>            │
 │  ├──────────────┼──────────┼────────────┼────────────────────────┤
 │  │ custom_entry │ hello.o  │ global     │ .text(section#1) + 0x0 │
 │  ╰──────────────┴──────────┴────────────┴────────────────────────╯
 ┴

error: failed to find the entry point of the executable
caused by: symbol _start is undefined


